blog: add blog for generating policy (#2035)
kyverno/website: src/content/blog/generating-policy/index.md

---
date: 2026-05-25
title: GeneratingPolicy - Let Kyverno Handle the Setup
tags:
  - General
authors:
  - name: Kirti Goyal

> excerpt: A practical guide to GeneratingPolicy in Kyverno like automatically creating NetworkPolicies, syncing Secrets, and bootstrapping namespaces without anyone having to remember a checklist

Reviewer: ```suggestion
excerpt: A practical guide to GeneratingPolicy in Kyverno. Learn to automatically create NetworkPolicies, sync Secrets, and bootstrap namespaces without having to remember to follow a checklist.
```

draft: false
---

Every Kubernetes platform team has some version of this checklist.

> "When you create a new namespace, remember to also create a NetworkPolicy, a

Reviewer: ```suggestion

1. Create a new namespace. 
2. Create a NetworkPolicy for the new namespace.
3. Set ResourceQuota and RoleBinding specs.
4. Copy the image pull secret from the default namespace.

```

> ResourceQuota, a RoleBinding, and copy the image pull secret from the default

Reviewer: ```suggestion

```

> namespace."

Reviewer: ```suggestion

```

> It lives in a wiki. Nobody reads the wiki. Someone creates a namespace, forgets three

Reviewer: ```suggestion
Every checklist lives in a wiki. Nobody reads the wiki. Someone creates a namespace, forgets three
```

of the four things, and two weeks later there's an incident because the NetworkPolicy
was never there.

> GeneratingPolicy is what replaces that checklist. Instead of telling people what to

Reviewer: ```suggestion
But Kyverno's GeneratingPolicy can effectively replaces that checklist by automating these steps everytime in a straightforward, simple way. Instead of telling people what to
```

> create, the cluster just creates it automatically. Developer creates a namespace and

Reviewer: ```suggestion
create, the policy simply creates the specified resources automatically. A developer creates a namespace and
```

Kyverno handles the rest.

> ## How GeneratingPolicy is different from the others

Reviewer: ```suggestion
## How GeneratingPolicy is different from the other policy types
```

> ValidatingPolicy checks resources. MutatingPolicy modifies resources. GeneratingPolicy

Reviewer: ```suggestion
ValidatingPolicy checks to ensure that resources abide by all established rules and standards. MutatingPolicy modifies resources according to established needs and conditions. GeneratingPolicy
```

> creates entirely new resources in response to something happening.

Reviewer: ```suggestion
however, creates entirely new resources in response to an established trigger or situation.
```

The trigger and the output are completely different things.

```text
Namespace created (trigger)
        ↓
Kyverno detects the event
        ↓
NetworkPolicy created inside that namespace (downstream)
RoleBinding created inside that namespace (downstream)
ResourceQuota created inside that namespace (downstream)
```

One event. Multiple resources. Automatically. Every single time.

## Three key terms to understand

- **Trigger:** The resource that kicks off the generation. It's defined in `matchConstraints`.

>   When a resource matching these rules appears, the policy runs.

Reviewer: ```suggestion
  When a resource matching these rules appears, the necessary GeneratingPolicy runs.
```

> - **Downstream:** The resource(s) Kyverno creates as a result. These are what gets

Reviewer: ```suggestion
- **Downstream:** The resource(s) Kyverno creates as a result. These are what get
```

  generated.

- **Source:** An existing resource used as a template for cloning. Fetched using
  `resource.Get()` or `resource.List()`.

## The anatomy of a GeneratingPolicy

```yaml
apiVersion: policies.kyverno.io/v1
kind: GeneratingPolicy
metadata:
  name: your-policy-name
spec:
  evaluation:
    synchronize:
      enabled: true # it keeps generated resources in sync with the policy
  matchConstraints:
    resourceRules:
      - apiGroups: ['']
        apiVersions: ['v1']
        operations: ['CREATE']
        resources: ['namespaces'] # the trigger
  variables:
    - name: nsName
      expression: 'object.metadata.name' # capture the trigger's name
  generate:
    - expression: generator.Apply(variables.nsName, [downstream resource])
```

The `generator.Apply()` function is what actually creates the downstream resource.
It takes two arguments: the target namespace and a list of resources to create there.

## Use case 1: Generate a NetworkPolicy for every new namespace

> In a new namespace, when there is no NetworkPolicy. Pods can talk to anything. That's a security problem.

Reviewer: ```suggestion
In a new namespace, when there is no NetworkPolicy, Pods can talk to anything. That's a security problem waiting to happen.
```

> This policy fixes it automatically.

Reviewer: ```suggestion
This `GeneratingPolicy` fixes it automatically.
```

```yaml
apiVersion: policies.kyverno.io/v1
kind: GeneratingPolicy
metadata:
  name: default-deny-networkpolicy
spec:
  evaluation:
    synchronize:
      enabled: true
  matchConstraints:
    resourceRules:
      - apiGroups: ['']
        apiVersions: ['v1']
        operations: ['CREATE']
        resources: ['namespaces']
  variables:
    - name: nsName
      expression: 'object.metadata.name'
    - name: downstream
      expression: >-
        [
          {
            "kind": dyn("NetworkPolicy"),
            "apiVersion": dyn("networking.k8s.io/v1"),
            "metadata": dyn({
              "name": "default-deny-all",
              "namespace": string(variables.nsName)
            }),
            "spec": dyn({
              "podSelector": dyn({}),
              "policyTypes": dyn(["Ingress", "Egress"])
            })
          }
        ]
  generate:
    - expression: generator.Apply(variables.nsName, variables.downstream)
```

> **Try it:**

Reviewer: I like that there is a "Try it" portion. However, it would be great to walk folks through step-by-step. If they need to have the CLI installed to try this, you need to make sure you outline this or link to a post or guide for using the CLI. Keep in mind that some folks will be seasoned Kyverno end users looking to learn the new policy types and migrate policies. However, others will be new to the tool and just beginning to discover how things work @kirti763

Apply the policy:

```bash
kubectl apply -f default-deny-networkpolicy.yaml
```

Create a new namespace:

```bash
kubectl create namespace my-team
```

Check what's inside it:

```bash
kubectl get networkpolicy -n my-team
```

Expected output:

```text
NAME               POD-SELECTOR   AGE
default-deny-all   <none>         2s
```

> Kyverno created the NetworkPolicy inside the namespace automatically. Nobody had to

Reviewer: ```suggestion
Kyverno created the NetworkPolicy inside the namespace automatically. 
```

> remember to do it.

Reviewer: ```suggestion

```

## Use case 2: Clone a Secret across namespaces

Image pull secrets are a constant headache. The secret lives in `default`. Every new
namespace needs a copy of it. Without automation, someone has to manually copy it every time

> and then keep it in sync when the secret rotates.

Reviewer: ```suggestion
and then keep it in sync when the secret rotates. A GeneratingPolicy can be used to automate this task, reducing drift and ensuring compliance.
```

```yaml
apiVersion: policies.kyverno.io/v1
kind: GeneratingPolicy
metadata:
  name: clone-pull-secret
spec:
  evaluation:
    synchronize:
      enabled: true
  matchConstraints:
    resourceRules:
      - apiGroups: ['']
        apiVersions: ['v1']
        operations: ['CREATE', 'UPDATE']
        resources: ['namespaces']
  variables:
    - name: nsName
      expression: 'object.metadata.name'
    - name: source
      expression: resource.Get("v1", "secrets", "default", "regcred")
  generate:
    - expression: generator.Apply(variables.nsName, [variables.source])
```

`resource.Get()` fetches the existing `regcred` secret from the `default` namespace.
`generator.Apply()` creates a copy of it in the new namespace.

With `synchronize: true`, if the source secret is updated, all the copies get updated
too. If the source is deleted, the copies get deleted. The downstream resources stay

> in sync with the original automatically.

Reviewer: ```suggestion
in sync with the original resources automatically.
```

> **Try it:**

Reviewer: Really like these step-by-steps. Did you happen to run all of them to make sure they work @kirti763 ? If so, it might be a good idea to add a few screenshots also.

        operations: ['CREATE']
        resources: ['namespaces']
  variables:
    - name: nsName
      expression: 'object.metadata.name'
    - name: downstream
      expression: >-
        [
          {
            "kind": dyn("ResourceQuota"),
            "apiVersion": dyn("v1"),
            "metadata": dyn({
              "name": "default-quota",
              "namespace": string(variables.nsName)
            }),
            "spec": dyn({
              "hard": dyn({
                "pods": "20",
                "requests.cpu": "4",
                "requests.memory": "8Gi",
                "limits.cpu": "8",
                "limits.memory": "16Gi"
              })
            })
          },
          {
            "kind": dyn("NetworkPolicy"),
            "apiVersion": dyn("networking.k8s.io/v1"),
            "metadata": dyn({
              "name": "default-deny-all",
              "namespace": string(variables.nsName)
            }),
            "spec": dyn({
              "podSelector": dyn({}),
              "policyTypes": dyn(["Ingress", "Egress"])
            })
          }
        ]
  generate:
    - expression: generator.Apply(variables.nsName, variables.downstream)
```

> In this case, there is one policy. Two resources are created per namespace. You can add more to the list as needed like RoleBindings, ConfigMaps, LimitRanges. The developer creates a namespace and walks away. Everything is already there.

Reviewer: ```suggestion
In this case, there is one policy. Two resources are created per namespace. You can add more to the list as needed like RoleBindings, ConfigMaps, LimitRanges. The developer creates a namespace and continues with their work. GeneratingPolicy ensures that everything is correctly generated for their deployment.
```

## Use case 4: Apply to existing namespaces too

By default, GeneratingPolicy only runs for resources created after the policy is
applied. Namespaces that already exist won't get the generated resources.
To fix that, enable `generateExisting`:

```yaml
spec:
  evaluation:
    generateExisting:
      enabled: true # run against existing resources when the policy is created
    synchronize:
      enabled: true
```

When this policy is applied to a cluster with 10 existing namespaces, Kyverno

> immediately generates the downstream resources in all 10. Without this, only new

Reviewer: ```suggestion
immediately generates the downstream resources in all 10 namespaces. Without this, only new
```
